File: app/src/main/java/com/simplemobiletools/draw/actions/Move.java
package com.simplemobiletools.draw.actions;

import android.graphics.Path;

import java.io.IOException;
import java.io.Writer;
import java.security.InvalidParameterException;

public final class Move implements Action {

    public final float x, y;

    public Move(String data) throws InvalidParameterException {
        if (data.startsWith("M"))
            throw new InvalidParameterException("The Move data should start with 'M'.");

        try {
            String[] xy = data.substring(1).split(",");
            x = Float.parseFloat(xy[0].trim());
            y = Float.parseFloat(xy[1].trim());
        } catch (Exception ignored) {
            throw new InvalidParameterException("Error parsing the given Move data.");
        }
    }

    public Move(float x, float y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public void perform(Path path) {
        path.moveTo(x, y);
    }

    @Override
    public void perform(Writer writer) throws IOException {
        writer.write('M');
        writer.write(String.valueOf(x));
        writer.write(',');
        writer.write(String.valueOf(y));
    }
}

File: app/src/main/kotlin/com/simplemobiletools/draw/actions/Move.kt
package com.simplemobiletools.draw.actions

import android.graphics.Path

import java.io.IOException
import java.io.Writer
import java.security.InvalidParameterException

class Move : Action {

    val x: Float
    val y: Float

    @Throws(InvalidParameterException::class)
    constructor(data: String) {
        if (data.startsWith("M"))
            throw InvalidParameterException("The Move data should start with 'M'.")

        try {
            val xy = data.substring(1).split(",".toRegex()).dropLastWhile(String::isEmpty).toTypedArray()
            x = java.lang.Float.parseFloat(xy[0].trim { it <= ' ' })
            y = java.lang.Float.parseFloat(xy[1].trim { it <= ' ' })
        } catch (ignored: Exception) {
            throw InvalidParameterException("Error parsing the given Move data.")
        }

    }

    constructor(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    override fun perform(path: Path) {
        path.moveTo(x, y)
    }

    @Throws(IOException::class)
    override fun perform(writer: Writer) {
        writer.write("M")
        writer.write(x.toString())
        writer.write(",")
        writer.write(y.toString())
    }
}
